EIPS/eip-9999.md

```diff
@@ -0,0 +1,92 @@
+
+## Motivation
+
+Currently, transactions without an explicit transaction access list cannot be efficiently parallelized, as the execution engine cannot determine in advance which addresses and storage slots will be accessed. While transaction-level access lists exist, they are not enforced, making it difficult to optimize execution pipelines.
+
+By enforcing access lists at the block level, we shift the responsibility of declaring storage accesses to the block builder. This enables the execution engine to efficiently parallelize transaction execution, knowing in advance the exact scope of storage interactions for each transaction. 
+
+With access lists at the block level, transactions can be scheduled and executed in parallel, allowing for performance improvements.
+
+## Specification
+
+### Block Header Modification
+
+Introduce a new field in the block header:
+
+```python
+class BlockBody:
+    ...
+    # Each access → ordered list of tx indices and associated values
+    block_level_accesses: Dict[
+        Tuple[Address, StorageKey],
+        Tuple[
+            List[int],  # The transaction indices that accessed this slot
+            List[Union[Value, Tuple[Value, Optional[Value]]]]  # The corresponding read or write values
+        ]
+    ]
+```
+
+The block_level_access_list maps each accessed address and storage key to the index of the transaction that accesses it, along with the corresponding storage value. This structure enables efficient validation and execution by explicitly tying storage interactions and their values to specific transactions.
+
+### State Transition Function
+
+Modify the state transition function to validate the block-level access lists asynchronously:
+
+```python 
+def state_transition(block: Block) -> None:
+    computed_access_list = {}
+
+    for idx, tx in enumerate(block.transactions):
+        accessed_items = execute_transaction(tx)  # returns {(address, slot): (value_before, optional_value_after)}
+        for key, val in accessed_items.items():
+            computed_access_list[key] = (idx, val)
+
+    if computed_access_list != block.header.block_level_access_list:
+        raise InvalidBlock("Mismatch in block-level access list or values.")
+
+```
+
+Transactions must not read or write to any address or storage key not declared in the block-level access list. Additionally, the block is invalid if the access list contains entries for addresses or storage keys that were not actually accessed, or if the recorded values deviate from those observed during execution.
+
+## Rationale
+
+### Block Size Considerations
+
+Including access lists increases block size, potentially impacting network propagation times. Empirically, typical access lists are around 30 KiB, with occasional peaks up to 120 KiB. Efficient encoding and potential limits on access list sizes should be considered.
```

Comment:
Block headers are supposed to be small. A hash of the list (either a MPT trie or whatever trie and take the root hash of this) should be the value in the header and the actual data should be in the block body (like transactions or withdrawals)

Reply:
Yeah agree with that. Let's put the list in the body and a commitment in the header.